src/Storages/HDFS/AsynchronousReadBufferFromHDFS.cpp

```diff
@@ -0,0 +1,230 @@
+#include "AsynchronousReadBufferFromHDFS.h"
+
+#if USE_HDFS
+#include <mutex>
+#include <base/logger_useful.h>
+#include <Storages/HDFS/HDFSCommon.h>
+#include <Storages/HDFS/ReadBufferFromHDFS.h>
+#include <Disks/IO/ThreadPoolRemoteFSReader.h>
+
+namespace CurrentMetrics
+{
+    extern const Metric AsynchronousReadWait;
+}
+
+namespace ProfileEvents
+{
+    extern const Event AsynchronousReadWaitMicroseconds;
+    extern const Event RemoteFSSeeks;
+    extern const Event RemoteFSPrefetches;
+    extern const Event RemoteFSCancelledPrefetches;
+    extern const Event RemoteFSUnusedPrefetches;
+    extern const Event RemoteFSPrefetchedReads;
+    extern const Event RemoteFSUnprefetchedReads;
+    extern const Event RemoteFSBuffers;
+}
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int CANNOT_SEEK_THROUGH_FILE;
+    extern const int SEEK_POSITION_OUT_OF_BOUND;
+    extern const int LOGICAL_ERROR;
+}
+
+AsynchronousReadBufferFromHDFS::AsynchronousReadBufferFromHDFS(
```

Comment:
`AsynchronousReadIndirectBufferFromRemoteFS` is actually needed for MergeTree Engine when reading from remote filesystems, and `ReadBufferFromRemoteFSGather` is hard coded in its implementation. So I create `AsynchronousReadBufferFromHDFS` for Hive Engine.

Do you think it is necessary to refactor `AsynchronousReadIndirectBufferFromRemoteFS` to make it more universal, not only can be used by MergeTree Engine, but also Hive Engine?

Reply:
> Do you think it is necessary to refactor AsynchronousReadIndirectBufferFromRemoteFS to make it more universal, not only can be used by MergeTree Engine, but also Hive Engine?

1. For sure it is not ok to leave this `AsynchronousReadBufferFromHDFS` as it is now, because it is 95% copy-paste of `AsynchronousReadIndirectBufferFromRemoteFS`.
2. Actually I think for just storages it is better to use `ParallelReadBuffer` and not to use this buffer with prefetching. And it is anyway not possible to use both implementations, but the `ParallelReadBuffer` version should be better.

Reply:
I'm just curious why you don't advice to use prefetching for Hive. Because prefetching better to be used in disk layer? Or because it's performance benefit is litter compared to `ParallelReadBuffer`?

Reply:
> Or because it's performance benefit is litter compared to ParallelReadBuffer?

I did not actually test performance benefit of it compared to `ParallelReadBuffer`, it would be good if you did that in this PR btw. 

> Because prefetching better to be used in disk layer? 

Yes, it was designed for disk layer. And because there will be now inconsistency for storages -- somewhere ParallelReadBuffer is used, somewhere prefetching.
I would rather have the same method everywhere.

Let's implement `ParallelReadBuffer` in this PR and see the performance difference. If prefetching buffer has some benefit compared to `ParalllelReadBuffer`, we'll leave it. At the moment I am very doubtful about having two different implementations.

My opinion: your buffer will be better for comparatively small-sized files. For larger files `ParallelReadBuffer` will be better.

Reply:
Since https://github.com/ClickHouse/ClickHouse/pull/36964 didn't prove obvious performance benefit for HDFS, pls reconsider this PR again. Thanks!